projects/packages/masterbar/tests/php/test-class-base-admin-menu.php

```diff
@@ -0,0 +1,185 @@
+	}
+
+	/**
+	 * Returning the environment into its initial state.
+	 *
+	 * @after
+	 */
+	public function tear_down() {
+		WorDBless_Options::init()->clear_options();
+		WorDBless_Users::init()->clear_all_users();
+	}
+
+	/**
+	 * Test get_instance.
+	 *
+	 * @covers ::get_instance
+	 * @covers ::__construct
+	 */
+	public function test_get_instance() {
+
+		$admin_menu = $this->get_concrete_menu_admin();
+
+		$instance = $admin_menu::get_instance();
+
+		$this->assertInstanceOf( Base_Admin_Menu::class, $instance );
+		$this->assertSame( $instance, static::$admin_menu );
+
+		$this->assertSame( 99998, has_action( 'admin_menu', array( $instance, 'reregister_menu_items' ) ) );
+		$this->assertSame( 11, has_action( 'admin_enqueue_scripts', array( $instance, 'enqueue_scripts' ) ) );
+	}
+
+	/**
+	 * Tests add_admin_menu_separator
+	 *
+	 * @covers ::add_admin_menu_separator
+	 */
+	public function test_add_admin_menu_separator() {
+		global $menu;
+
+		// Start with a clean slate.
+		$temp_menu = $menu;
+		$menu      = array();
+
+		static::$admin_menu->add_admin_menu_separator( 15 );
+		static::$admin_menu->add_admin_menu_separator( 10, 'manage_options' );
+		'@phan-var non-empty-array $menu';
+		$this->assertSame( 'manage_options', $menu[10][1] );
+		$this->assertStringContainsString( 'separator-custom-', $menu[10][2] );
+		$this->assertSame( 'read', $menu[15][1] );
+		$this->assertStringContainsString( 'separator-custom-', $menu[15][2] );
+
+		// Restore filtered $menu.
+		$menu = $temp_menu;
+	}
+
+	/**
+	 * Tests preferred_view
+	 *
+	 * @covers ::set_preferred_view
+	 * @covers ::get_preferred_views
+	 * @covers ::get_preferred_view
+	 */
+	public function test_preferred_view() {
+		$this->assertSame( 'default', static::$admin_menu->get_preferred_view( 'test.php' ) );
+		$this->assertSame( 'unknown', static::$admin_menu->get_preferred_view( 'test.php', false ) );
+
+		update_user_option( get_current_user_id(), 'jetpack_admin_menu_link_destination', true );
+		$this->assertSame( 'classic', static::$admin_menu->get_preferred_view( 'test.php' ) );
+		delete_user_option( get_current_user_id(), 'jetpack_admin_menu_link_destination' );
+
+		static::$admin_menu->set_preferred_view( 'test.php', 'classic' );
+		$this->assertSame( 'classic', static::$admin_menu->get_preferred_view( 'test.php' ) );
+
+		static::$admin_menu->set_preferred_view( 'test.php', 'default' );
+		$this->assertSame( 'default', static::$admin_menu->get_preferred_view( 'test.php', false ) );
+	}
+
+	/**
+	 * Tests preferred_view
+	 *
+	 * @covers ::handle_preferred_view
+	 */
+	public function test_handle_preferred_view() {
+		// @see p9dueE-3LL-p2#comment-6669
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			$this->markTestSkipped( 'Does not work on WP.com as handle_preferred_view() performs a redirect and then terminates the execution.' );
+		}
+
+		$this->markTestSkipped( 'Temporarily skip this test till we update the package with ExitException requirement.' );
```

Comment:
The file `projects/plugins/jetpack/tests/php/redefine-exit.php` is relevant to this. It uses [Patchwork](https://github.com/antecedent/patchwork/) to "redefine" `exit()` and `die()` to throw catchable exceptions instead of actually exiting so PHPUnit can properly fail relevant tests instead of erroneously exiting in the middle.

Probably we should clean that up a little and make it a separate package, which we'd then be able to use in this package's tests. I'm willing to do that if you want.

Reply:
That was exactly what I had in mind to suggest as I didn't want to copy the file in the package and add Patchwork. I'd really appreciate it if you could provide the package but it's definitely not a blocker for the work here so you can prioritize this per your schedule. Many thanks!

Reply:
Using the package via https://github.com/Automattic/jetpack/pull/37342/commits/6c7b5837b4f33d8f62b45bcbdb2d0d44444737bd

Many thanks for adding it so promptly!